### PR TITLE
Abstract hook registration and handle passive and dispatch

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Abstract hook registration and handle passive and dispatch

Resolves: #263 

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: netebpfext/ebpf_ext_attach_provider.c [hunk 1]
```diff
@@ -0,0 +1,185 @@
+/*
+ *  Copyright (c) Microsoft Corporation
+ *  SPDX-License-Identifier: MIT
+ */
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.c:4]:
```suggestion
// Copyright (c) Microsoft Corporation
// SPDX-License-Identifier: MIT
```
Please use the above style per https://github.com/microsoft/ebpf-for-windows/blob/master/docs/DevelopmentGuide.md#license-header

Same comment on other files

File: netebpfext/ebpf_ext_attach_provider.c [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    };
+} ebpf_ext_attach_hook_provider_registration_t;
+
+static ebpf_result_t
+_ebpf_ext_attach_provider_client_attach_callback(
+    void* context,
+    const GUID* client_id,
+    void* client_binding_context,
+    const ebpf_extension_data_t* client_data,
+    const ebpf_extension_dispatch_table_t* client_dispatch_table)
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.c:37]: SAL?  Same question on line 105 etc

File: netebpfext/ebpf_ext_attach_provider.c [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+ebpf_result_t
+ebpf_ext_attach_register_provider(
+    const ebpf_attach_type_t* attach_type,
+    ebpf_ext_hook_execution_t execution_type,
+    ebpf_ext_attach_hook_provider_registration_t** registration)
+{
+    ebpf_result_t return_value;
+    ebpf_ext_attach_hook_provider_registration_t* local_registration = NULL;
+
+    local_registration = ebpf_allocate(sizeof(ebpf_ext_attach_hook_provider_registration_t));
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.c:144]: check for alloc failure

File: netebpfext/ebpf_ext_attach_provider.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
+/*
+ *  Copyright (c) Microsoft Corporation
+ *  SPDX-License-Identifier: MIT
+ */
+
+#pragma once
+
+#include <ntddk.h>
+
+#include <ebpf_platform.h>
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.h:10]:
```suggestion
#include "ebpf_platform.h"
```
Only use <> for system headers

File: netebpfext/ebpf_ext_attach_provider.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+} ebpf_ext_hook_execution_t;
+
+typedef struct _ebpf_ext_attach_hook_provider_registration ebpf_ext_attach_hook_provider_registration_t;
+
+/**
+ * @brief Protect the registration from rundown.
+ *
+ * @param[in] registration Registration to protect.
+ * @retval true - The caller should proceed.
+ * @retval false - Rundown has occurred.
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.h:25]:
```suggestion
 * @retval true The caller should proceed.
 * @retval false Rundown has occurred.
```

File: netebpfext/ebpf_ext_attach_provider.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ */
+bool
+ebpf_ext_attach_enter_rundown(_In_ ebpf_ext_attach_hook_provider_registration_t* registration);
+
+/**
+ * @brief Unprotect the registration from rundown.
+ *
+ * @param registration
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.h:33]: add [in,out] style annotation

File: netebpfext/ebpf_ext_attach_provider.h [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * @param[in] execution_type Execution type for the hook (passive or dispatch).
+ * @retval EBPF_SUCCESS The operation was successful.
+ * @retval EBPF_NO_MEMORY Unable to allocate resources for this
+ * operation.
+ */
+ebpf_result_t
+ebpf_ext_attach_register_provider(
+    const ebpf_attach_type_t* attach_type,
+    ebpf_ext_hook_execution_t execution_type,
+    ebpf_ext_attach_hook_provider_registration_t** registration);
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.h:52]: Add SAL.  Same for line 60 etc

File: netebpfext/net_ebpf_ext.c [hunk 8]
```diff
@@ -613,123 +607,31 @@ _net_ebpf_ext_no_op_flow_delete(uint16_t layer_id, uint32_t fwpm_callout_id, uin
...
-        NULL,
-        &_ebpf_xdp_hook_provider_registration,
-        _net_ebpf_ext_provider_client_attach_callback,
-        _net_ebpf_ext_provider_client_detach_callback);
-
+    return_value = ebpf_ext_attach_register_provider(
+        &EBPF_ATTACH_TYPE_XDP, EBPF_EXT_HOOK_EXECUTION_DISPATCH, &_ebpf_xdp_hook_provider_registration);
     if (return_value != EBPF_SUCCESS) {
-        goto Done;
+        return STATUS_UNSUCCESSFUL;
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:617]: ebpfcore.sys has _ebpf_result_to_ntstatus(), should we export that for use by extensions like here?

File: netebpfext/ebpf_ext_attach_provider.c [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+#include "ebpf_ext_attach_provider.h"
+
+typedef struct _ebpf_ext_attach_hook_provider_registration
+{
+    ebpf_extension_data_t* provider_data;
+    ebpf_extension_provider_t* provider;
+    GUID client_id;
+    void* client_binding_context;
+    const ebpf_extension_data_t* client_data;
+    const ebpf_result_t (*invoke_hook)(void* bind_context, void* context, uint32_t* result);
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.c:13]: Add SAL to arguments

File: netebpfext/net_ebpf_ext.c (hunk 8 above)
Comment (Alan-Jowett) [reply netebpfext/net_ebpf_ext.c:615]: Issue #267 filed

File: netebpfext/ebpf_ext_attach_provider.c (hunk 2 above)
Comment (dthaler) [reply netebpfext/ebpf_ext_attach_provider.c:35]: Still missing SAL annotation here and in what is now 107

File: netebpfext/ebpf_ext_attach_provider.h [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ */
+void
+ebpf_ext_attach_leave_rundown(_In_ ebpf_ext_attach_hook_provider_registration_t* registration);
+
+/**
+ * @brief Register as a attach type provider.
+ *
+ * @param[in] attach_type Attach type to register for.
+ * @param[in, out] registration Registration to complete.
+ * @param[in] execution_type Execution type for the hook (passive or dispatch).
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.h:41]:
```suggestion
 * @param[in] execution_type Execution type for the hook (passive or dispatch).
 * @param[in, out] registration Registration to complete.
```
List params in the same order as in the prototype

File: netebpfext/ebpf_ext_attach_provider.h [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+ * @retval EBPF_SUCCESS The operation was successful.
+ * @retval EBPF_NO_MEMORY Unable to allocate resources for this
+ * operation.
+ */
+ebpf_result_t
+ebpf_ext_attach_register_provider(
+    _In_ const ebpf_attach_type_t* attach_type,
+    ebpf_ext_hook_execution_t execution_type,
+    _Outptr_ ebpf_ext_attach_hook_provider_registration_t** registration);
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.h:50]: The doxygen says this is [in,out] but here says only Out.  Fix whichever one is incorrect

File: netebpfext/ebpf_ext_attach_provider.c [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    } else {
+        // Queue a DPC to each CPU and wait for it to run.
+        // After it has run on each CPU we can be sure that no
+        // DPC is busy processing a hook.
+        uint32_t maximum_processor = KeQueryMaximumProcessorCount();
+        uint32_t processor;
+        for (processor = 0; processor < maximum_processor; processor++) {
+            KeSetTargetProcessorDpc(&registration->dispatch.rundown_dpc, (uint8_t)processor);
+            if (KeInsertQueueDpc(&registration->dispatch.rundown_dpc, NULL, NULL)) {
+                KeWaitForSingleObject(&registration->dispatch.rundown_wait, Executive, KernelMode, FALSE, NULL);
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply netebpfext/ebpf_ext_attach_provider.c:126]: Filed #270 to track this feedback.

**Review by shankarseal — APPROVED** [review]

**Pull request merged**